### DISCUSSION TITLE
OnUnsubscribe?

### DISCUSSION THREAD

**Discussion by JakeCooper** [enisdenjo/graphql-ws#283] [Q&A] [answered]
### Story

As a enjoyer of this library

I want my server, ideally my async generator, to be notified when the cleanup function, returned from client.subscribe, is called

<img width="659" alt="CleanShot 2021-12-15 at 12 05 49@2x" src="https://user-images.githubusercontent.com/5499880/146257320-a44e678b-a42b-42b6-a676-94fa8a11c988.png">

So that I can close downstream connections to other services (in this case a GRPC stream)

### Acceptance criteria

P0: server is notified on unsubscribe
P1: handler is notified of unsubscribe

**Comment by enisdenjo** [comment, marked as answer]
As per the Observable pattern, a stream is stopped with either an "error" or a "complete" event. If you dispose of the `client.subscribe` subscription, or if the subscription is completed by the server, or the query/mutation is done - the `complete` callback in the subscribed sink will be called.

On the server side, what you're looking for is therefore the [`onComplete` server hook](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#oncomplete). It is called for each query/mutation/subscription after it has completed for whatever reason (user disconnected, server completed, client completed/unsubscribed). An same on the error side with the [`onError` hook](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#onerror).

Additionally, if you're looking for hooks for when the client completely disconnects/closes the WebSocket connection, you might want to check [`onDisconnect`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#ondisconnect) and [`onClose`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#onclose) respectively.

**Reply by JakeCooper** [reply]
What we're seeing is that the onComplete server hook isn't being called until we try to yield from the server, at which point I suppose the library realizes that should be completed and then it resolves

However, if you're not yielding anything from the async generator, there's no way to know something is completed without attempting to yield

**Reply by JakeCooper** [reply]
I suppose, if we can agree that onComplete should be called when the cleanup function is called, I have a bug report which is that it doesn't seem to be called until something is yielded from the generator

Code Sample:

```
 subscribe: async (_, { deploymentId }) => {
    return (async function* () {
      // eslint-disable-next-line no-constant-condition
      while (true) {
        await new Promise(resolve => setTimeout(resolve, 1000));
        console.log("TICK");
        // yield "foo"; // onClose+onComplete will only be called if this is uncommented
      }
    })();
  }
```

Thank you for the lovely library BTW; it's been a [joy to hook up to Nexus and GRPC](https://twitter.com/JustJake/status/1469066409569402883)

**Reply by enisdenjo** [reply]
That seems to be a problem with something else. Which other libraries do you use for your streaming?

The `onComplete` is simply called when the underlying iterable returns, there's no magic there. If the iterable needs poking, the callback will be stalled as an effect of that.

**Reply by JakeCooper** [reply]
I'll provide both the client and server implementation for you. It's graphql-ws client and server so I think it should be fine, so long as apollo isn't messing something up major here.


client.ts
```
import { createClient } from "graphql-ws";
import {
  DeployLogsDocument,
  DeployLogsSubscriptionVariables,
  Subscription,
} from "../graphql/sdk";
import { useEffect, useState } from "react";
import { print } from "graphql";
import { backboardBaseUrl } from "../constants/domains";

const client = createClient({
  url: `${backboardBaseUrl.replace("https", "wss")}/graphql`,
});

export type RealtimeLogsResponse = Pick<Subscription, "deployLogs">;
export interface SubscriptionResponse<Response> {
  data: Response[];
  error: Error | null;
  isLoading: boolean;
  isClosed: boolean;
}

export const useDeploymentLogSubscription = (
  variables: DeployLogsSubscriptionVariables,
): SubscriptionResponse<RealtimeLogsResponse> => {
  return useSubscription<RealtimeLogsResponse>(DeployLogsDocument, variables);
};

const useSubscription = <T extends unknown>(
  query,
  variables,
): SubscriptionResponse<T> => {
  const [state, setState] = useState<SubscriptionResponse<T>>({
    data: [],
    isLoading: true,
    error: null,
    isClosed: false,
  });

  useEffect(() => {
    const onNext = ({ data: result }) => {
      // Dirty lil hack
      // Since the first field is always the same, use it to flatten query
      const queryName = Object.keys(result)[0];
      setState(state => ({
        ...state,
        isLoading: false,
        data: [...state.data, result[queryName]],
      }));
    };

    return client.subscribe(
      {
        query: print(query),
        variables: variables,
      },
      {
        next: onNext,
        error: err => setState({ ...state, error: err as Error }),
        complete: () => setState({ ...state, isClosed: true }),
      },
    );
  }, []);

  return state;
};
```

server.ts
```
import { ApolloServer } from "apollo-server-koa";
import { applyMiddleware } from "graphql-middleware";
import Koa from "koa";
import { logger } from "../clients/logger";
import {
  createContext,
  GQLContext,
  koaContextToApolloContext,
} from "../graphql/context";
import { permissions } from "../graphql/permissions";
import { schema } from "../graphql/schema";
import { errorHandler, shouldGQLErrorBeReturned } from "../middleware/graphql";
import { sentryTracing } from "../middleware/sentryTracing";
import { Server } from "ws";
import { useServer } from "graphql-ws/lib/use/ws";
import * as http from "http";

const gqlLogger = logger.child({ layer: "graphql" });

export const connectGraphQL = (httpServer: http.Server, app: Koa) => {
  const schemaWithMiddleware = applyMiddleware(schema, permissions);

  const server = new ApolloServer({
    schema: schemaWithMiddleware,
    logger: gqlLogger,
    context: koaContextToApolloContext,
    plugins: [errorHandler, sentryTracing],
    formatError: error => {
      return shouldGQLErrorBeReturned(error)
        ? error
        : new Error("Problem processing request");
    },
  });

  const wsServer = new Server({
    server: httpServer,
    path: "/graphql",
  });

  useServer(
    {
      schema,
      context: ctx => ctx.extra.context,
      onConnect: async (ctx) => {
        const gqlContext: GQLContext = await createContext({
          httpRequest: ctx.extra.request,
        });
        if (typeof ctx.extra.context != "object") {
          (ctx.extra.context as any) = {};
        }
        Object.assign(ctx.extra.context, gqlContext);
      },
      onComplete: async () => {
        console.log("DONE!") // Not called :(
      }
    },
    wsServer,
  );

  server.applyMiddleware({ app });
};
```

**Reply by enisdenjo** [reply]
I need the part where you do the streaming of subscriptions. So, the schema and its pubsub system.

**Reply by enisdenjo** [reply]
Apollo's `graphql-subscriptions` lib is nutorious for not returning iterables properly. 😉

**Reply by JakeCooper** [reply]
Not sure I follow:

This is where the action subscriptionField is defined, if that's what you're after?

```
export const deployLogs = subscriptionField("deployLogs", {
  type: "String",
  description: "Tail a stream of deploy logs",
  args: {
    projectId: nonNull(idArg()),
    deploymentId: nonNull(idArg()),
  },
  subscribe: async (root, { deploymentId }, ctx, info) => {
    return (async function* () {
      // Get the containerId from the deployment
      const deployment = await getDeploymentById(deploymentId);
      const containerId = deployment.containerId;
      if (containerId == null) {
        return "No logs available";
      }

      // Create a stream of the current logs
      const stream = streamDeployLogs({ containerId });

      for await (const response of stream) {
        yield response.Logs;
      }
    })();
  },
  resolve: (payload: string) => payload,
});
```

**Reply by JakeCooper** [reply]
I've heard that hence why I spun up the ws* server :S

I know in Apollo 3 they've distanced themselves entirely from subscriptions

**Reply by enisdenjo** [reply]
I bet the stream here:
```
const stream = streamDeployLogs({ containerId });
```
is not disposed properly, it's the one that stalls. How do you even propagate the `return` from the parent iterator to the stream itself?

Check out the ["client usage with AsyncIterator" recipe](https://github.com/enisdenjo/graphql-ws#async-iterator) for how to set up an iterator properly.

**Reply by JakeCooper** [reply]
I can give that a go and circle back. However worth clarifying: that streamDeployLogs function is on the server, which gets closed via a context cancel when the infrastructure closes the logstream, so I'm 99.9999% certain that's fine

An await on that stream shouldn't block your await on socket info because well, async JS right?

I'll attempt the above and circle back

**Reply by enisdenjo** [reply]
> An await on that stream shouldn't block your await on socket info because well, async JS right?

Of course it blocks, as long as the yield loop is alive - the subscription is alive.

The iterator from the `subscribe` in the schema **must** return for the subscription to be released.
```ts
export const deployLogs = subscriptionField("deployLogs", {
  type: "String",
  description: "Tail a stream of deploy logs",
  args: {
    projectId: nonNull(idArg()),
    deploymentId: nonNull(idArg()),
  },
  subscribe: async (root, { deploymentId }, ctx, info) => {
    //     👇 this one gets disposed/returned on subscription complete
    return (async function* () {
      // ...

      //   👇 this stream must also get disposed/returned
      const stream = streamDeployLogs({ containerId });

      // 👇 to break this loop and complete the subscription/iterator
      for await (const response of stream) {
        yield response.Logs;
      }
    })();
  },
  resolve: (payload: string) => payload,
});
```

**Reply by gschier** [reply]
_Thanks for the continued amazing support on this thread!_


> The iterator from the subscribe in the schema must return for the subscription to be released.

So, I guess my question is... if we have our GRPC stream listener inside our generator that never (or in a long time) returns, how can we detect that the WS client has stopped listening from within the generator in order to shut down the GRPC stream and exit the loop? It seems as though we can only exit after another response from the stream has arrived and we finished the next iteration.

The only thing I can think to do for now is to update our `streamDeployLogs` to send empty messages every `n` seconds in order to bump the loop and give it a chance to release the subscription. 

---

_I was able to put together a reproduction repo, if it helps: https://github.com/gschier/ws-demo_

**Reply by gschier** [reply]
I assume that, if my iterator never yields, it just gets stuck here on line 797?

![image](https://user-images.githubusercontent.com/587576/146285795-b3820dba-91b0-423a-b344-80510b38bef4.png)

**Reply by enisdenjo** [reply]
Exactly. Your iterator must yield and finally return/complete, for the subscription to behave as you expect and as it should.

> So, I guess my question is... if we have our GRPC stream listener inside our generator that never (or in a long time) returns, how can we detect that the WS client has stopped listening from within the generator in order to shut down the GRPC stream and exit the loop? It seems as though we can only exit after another response from the stream has arrived and we finished the next iteration.

When the WS client stops listening, the subscription iterator will have its `return` method called:
```ts
  // ...
  subscribe: async (root, { deploymentId }, ctx, info) => {
    //     👇 this one
    return (async function* () {
    // ...
```

It is your task to provide an appropriate `return` method for the subscription and propagate the invocation to your iterators, in this case to `streamDeployLogs`.

I'd recommend looking over the ["client usage with AsyncIterator" recipe](https://github.com/enisdenjo/graphql-ws#async-iterator) once more in-depth for how to do that properly.

**Reply by enisdenjo** [reply]
Also keep in mind that this is not _only_ about `graphql-ws` and the subscriptions working properly - if you don't dispose of the stream from `streamDeployLogs` properly, you're facing a memory leak.

**Reply by gschier** [reply]
> I'd recommend looking over the "client usage with AsyncIterator" recipe once more in-depth for how to do that properly.

Ah, yes that looks like it should work! I could only find the `next()` method documented for Async Iterators but just discovered that `throw()` and `return()` are part of the Generator spec instead. Thanks for pointing this out 👍🏻 

> Also keep in mind that this is not only about graphql-ws and the subscriptions working properly - if you don't dispose of the stream from streamDeployLogs properly, you're facing a memory leak.

Yep! We'll be sure to do this, and thanks again for being such a huge help! 🥳

**Reply by enisdenjo** [reply]
> I could only find the next() method documented for Async Iterators but just discovered that throw() and return() are part of the Generator spec instead.

Yes, you're right. _Iterators_ are objects which allow traversal over a (often) finite set of values that have an end; while _generators_, on the other hand, are actually functions that "generate" the next value on-demand. They are also iterators, but what makes them special is that they run only when `next` is invoked.

Because of these intrinsic differences - generators must implement the `throw` and `return` methods as means of communicating an end, while pure iterators don't have to.

**Comment by enisdenjo** [comment]
_Moved to https://github.com/enisdenjo/graphql-ws/discussions/283#discussioncomment-1817559._
